[MRG] ntp rework managment (#7)
bluegenes/2020-gep: content/05.project-management.md

## Workflow-Based Project Management

Computational project management is a learned skill that takes time to implement.
Biological analyses often span hundreds of steps and involve a myriad of decisions ranging from small-scale tool and parameter choices to larger-scale decisions around computational experimental design and statistical analyses.

> It is rare (if not impossible) to build or find a workflow that to analyze your data from start to finish without testing, troubleshooting, and iteration.

Reviewer: ```suggestion
It is rare (if not impossible) to build or find a workflow that analyzes your data from start to finish without testing, troubleshooting, and iteration.
```

Fortunately, with a little direction, workflow systems both simplify and improve computational project management.

> ### Systematically document your workflows

Reviewer: This paragraph doesn't really work as is...it doesn't set up for the information to come in the next few paragraphs, and doesn't provide strategies to combat the ways in which workflow self-documentation is limited.

The colors represent arms of the workflow that achieve a final result, such as a multiple sequence alignment of a protein of interest.
While the workflow is complex, it is coordinated by a workflow system, alleviating the need for a user to manage file interdependencies.](images/hu_dag.png){#fig:sgc_workflow}

### Version control your project

As your project develops, version control allows you to keep track of changes over time.
You may already do this in some ways, perhaps with frequent hard drive backups or by manually saving different versions of the same file  - e.g. by appending the date to a script name or appending "version_1" or "version_FINAL" to a manuscript draft.
For computational workflows, using version control systems such as Git or Mercurial can be used to properly keep track of all changes over time, even across multiple systems, scripting languages, and project contributors (see **Figure {@fig:version_control}**).
If a key piece of a workflow inexplicably stops working, good version control can allow you to rewind in time and identify differences from when the pipeline worked to when it stopped working.

> ![**Version Control** Version control systems (e.g. git, mercurial) work by storing incremental differences in files from one saved version ("commit") to the next. To visualize the differences between each version, text editors such as Atom and online services such as GitHub, GitLab and Bitbucket use red highlight to denote deletions, and green highlighting to denote additions. In this trivial example, a typo in version 1 (in red) was corrected (green). These systems are extremely useful for working code and manuscript development, as it is possible to return to the snapshot of any saved version. This means that version control systems save you from accidental deletions, preserve code you thought you no longer needed, and preserve a record of project change over time.](images/version_control.svg){#fig:version_control}

Reviewer: ```suggestion
![**Version Control** Version control systems (e.g. Git, Mercurial) work by storing incremental differences in files from one saved version ("commit") to the next. To visualize the differences between each version, text editors such as Atom and online services such as GitHub, GitLab and Bitbucket use red highlight to denote deletions, and green highlighting to denote additions. In this trivial example, a typo in version 1 (in red) was corrected (green). These systems are extremely useful for code and manuscript development, as it is possible to return to the snapshot of any saved version. This means that version control systems save you from accidental deletions, preserve code you thought you no longer needed, and preserve a record of project changes over time.](images/version_control.svg){#fig:version_control}
```

Shiny allows you to build interactive web pages using R code.
Data is manipulated  by R code in real-time in a web page, producing analysis and visualizations of a data set.
Shiny apps can contain user-specifiable parameters, allowing a user to control visualizations or analyses. As seen above, sample "PT1" is selected, and taxonomic ranks class and order are excluded.
Shiny apps allow collaborators who may or may not know R to change R visualisations to fit their interests.   
**B** Plotly heatmap of transcriptional profiling in human brain samples [@url:https://plotly.com/python/v3/ipython-notebooks/bioinformatics/#4-heatmap-of-gene-expression].
Hovering over a cell in the heatmap displays the sample names from the x and y axis, as well as the intensity value.
Plotting tools like plotly and vega-lite produce single interactive plots that can be shared with collaborators or integrated into websites [@url:https://plotly.com/; @doi:10.1109/TVCG.2016.2599030].
Interactive visualizations are also helpful in exploratory data analysis.
](images/interactive_viz.png){#fig:interactiveviz}

### Getting started developing workflows

> In our experience, the best way to have your workflow system work _for_ you is to include as much of your analysis as possible within the automated workflow framework, use self-documenting names, include analysis visualizations, and keep rigorous documentation alongside that would enable you to understand each decision and entirely reproduce any manual steps. Some of the tools discussed above will inevitably change over time, but these principles apply broadly and will help you design clear, well-documented, and reproducible analyses. Ultimately, you'll need to experiment with the ways that work for you -- what is most important is to develop a clear set of strategies and implement them tenaciously. Here are a few practical tips to get started:

Reviewer: ```suggestion
In our experience, the best way to have your workflow system work _for_ you is to include as much of your analysis as possible within the automated workflow framework, use self-documenting names, include analysis visualizations, and keep rigorous documentation alongside your workflow that enables you to understand each decision and entirely reproduce any manual steps. 
Some of the tools discussed above will inevitably change over time, but these principles apply broadly and will help you design clear, well-documented, and reproducible analyses. 
Ultimately, you'll need to experiment with the ways that work for you -- what is most important is to develop a clear set of strategies and implement them tenaciously. 
Here are a few practical tips to get started.
```

**Start with working code**
The best way to develop your own workflow analysis is to start from working code: either from the workflow tutorials or code sharing websites like GitHub, GitLab, and Bitbucket.
If a workflow is available through Binder, you can test and experiment with workflow modification on Binder's cloud system, without needing to install a workflow manager or software management tool.
When you're building a workflow for the first time, choose code in official repositories or reasonably well-used that provides sample data that can be run to verify the analysis works.
The "Workflows and Software Management" Section links to a number of official repositories containing tutorials and example biological analysis workflows.

**Test with subsampled data**

> After verifying your chosen example workflow is functional, try running it with your own data or some public data related to your species or condition of interest (see **Table {@tbl: seq_resources}**)

Author: ```suggestion
After verifying your chosen example workflow is functional, try running it with your own data or some public data related to your species or condition of interest (see **Table {@tbl:seq_resources}**)
```

For example, if working with FASTQ data, you can subsample the first million lines of a file (first 250k reads) by running:

`head -n 1000000 FASTQ_FILE.fq > test_fastq.fq`

While there are many more sophisticated ways to subsample reads, this technique should be sufficient for testing each step of a workflow prior to running your full dataset.

**Document your process** Write documentation in plain text, so your notes files can be saved under version control with your workflow. We recommend using Markdown language for your notes so they can include helpful headings, code formatting, and embedded images. We recommend using software with visual text previewing such as Hackmd or Atom [@url:https://hackmd.io/, @url:https://atom.io/].

**Develop your workflow**
Now that you have the workflow running on new data, you can begin to modify a workflow step to meet your needs.
Run each modified step on your subsampled data, checking that the output files look correct.
Keep the tool documentation and tutorials on hand; as with any language, remembering workflow-specific syntax takes time and practice.

**Back up early and often**

> Now that you've written new code, the first thing to do is back up your changes in an online repository such as GitHub, GitLab, or Bitbucket.

Reviewer: ```suggestion
As you write new code, back up your changes in an online repository such as GitHub, GitLab, or Bitbucket.
```

Workflow systems provide built-in tools to monitor resource usage for each step.
Turn on this reporting according to workflow documentation and record the resources required for running each tool on a full-size single sample.
Use these estimates to set appropriate resource limits for each step when executing the workflow on your remaining samples.

**Find a community and ask for help when you need it**
If there are local user groups for your workflow language, try to get involved with the community.
When you are first learning, help from more advanced users can save you hours of frustration.
After you've progressed, providing that same help to new users can help you cement the syntax in your mind and tackle more advanced uses.
If no local community is available, look online for forums related to your workflow language.
These workflow systems have been enthusiastically adopted by the open science community, and as a consequence, there is a critical mass of tutorials and open access code, code discussion on forums and via social media, particularly twitter.

> Be respectful of people's time and energy: post in the relevant workflow forums only when you've hit a stopping point you're unable to work through.

Reviewer: ```suggestion
Be respectful of people's time and energy: post in the relevant workflow forums only when you have hit a stopping point you are unable to work through.
```
